packages/bitcore-node/src/services/p2p.ts

```diff
@@ -98,13 +99,18 @@ export class P2pService {
         this.invCache.set(hash);
         this.events.emit(hash, message.block);
         if (!this.syncing) {
-          try {
-            await this.processBlock(block);
-            this.events.emit('block', message.block);
-          } catch (err) {
-            logger.error(`Error syncing ${chain} ${network}`, err);
-            this.sync();
-          }
+          await this.blockProcessLock;
```

Comment:
This was the problem! there was a race condition on `processBlock` that would throw two `reorg` errors where the second `reorg` error drops the tip that sync has constructed, causing multiple blocks to be lost.  

Ironically, I arrived at the same conclusion just after I posted the follow up comment on #1657. My fix (54ce5f7944c31fe72bc8a421c1f84fcb6ea777cf) used a lock inside of `this.processBlock(block)` the same way `this.sync()` is locked. 

Is this approach better than just locking processBlock from the inside?